### PR TITLE
lightning: allow caller to set a logger

### LINKED ISSUE #35556: lightning can be set a logger when used as a library
## Feature Request

**Is your feature request related to a problem? Please describe:**

When use lightning as a library and there are many lightning running, it's hard to tell their logs apart.

**Describe the feature you'd like:**

Caller can set labels to logger, and set the logger to lightning

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #35556

Problem Summary:

### What is changed and how it works?

don't use the global referenced logger, we now use logfer from RunOptions

### Check List

Tests 

- [ ] pass existing tests

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by lance6716** [comment]
CI passed 😭 PTAL @gozssky 

and please recommend another reviewer of lightning

File: br/pkg/lightning/lightning.go [hunk 1]
```diff
@@ -315,6 +317,15 @@ func (l *Lightning) RunOnceWithOptions(taskCtx context.Context, taskCfg *config.
 		opt(o)
 	}
 
+	if o.logger != nil {
+		loggerBackup := l.logger
+		// restore the logger to server's logger, after this task is finished.
+		defer func() {
+			l.logger = loggerBackup
+		}()
+		l.logger = log.Logger{Logger: o.logger}
```
Comment (gozssky) [inline br/pkg/lightning/lightning.go:326]: Please always pass logger in `option` and don't use and change `l.logger`.

**Review by gozssky — COMMENTED** [review]
rest LGTM

File: br/pkg/lightning/backend/kv/sql2kv.go [hunk 2]
```diff
@@ -267,7 +272,7 @@ func logKVConvertFailed(logger log.Logger, row []types.Datum, j int, colInfo *mo
 		log.ShortError(err),
 	)
 
-	log.L().Error("failed to covert kv value", logutil.RedactAny("origVal", original.GetValue()),
+	logger.Error("failed to covert kv value", logutil.RedactAny("origVal", original.GetValue()),
```
Comment (buchuitoudegou) [inline br/pkg/lightning/backend/kv/sql2kv.go:275]:
```suggestion
	logger.Error("failed to convert kv value", logutil.RedactAny("origVal", original.GetValue()),
```

Seems a typo. Would you mind fixing it in this PR?

File: br/pkg/lightning/lightning.go [hunk 3]
```diff
@@ -357,7 +368,7 @@ var (
 
 func (l *Lightning) run(taskCtx context.Context, taskCfg *config.Config, o *options) (err error) {
 	build.LogInfo(build.Lightning)
-	log.L().Info("cfg", zap.Stringer("cfg", taskCfg))
+	l.logger.Info("cfg", zap.Stringer("cfg", taskCfg))
```
[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/lightning.go]: https://github.com/pingcap/tidb/pull/35528/commits/a5b04631a492675120fca6ab15b4b0672cce351a

not sure I understood it correctly

File: br/pkg/lightning/backend/kv/sql2kv.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/kv/sql2kv.go]: done

**Review by gozssky — APPROVED** [review]

File: br/pkg/lightning/backend/kv/sql2kv.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/kv/sql2kv.go]: PTAL @buchuitoudegou

**Review by buchuitoudegou — APPROVED** [review]

**Pull request merged**